dsgrid/utils/dataset.py

```diff
@@ -0,0 +1,117 @@
+    # After remapping, rows in load_data_lookup for standard_handler and rows in load_data for
+    # one_table_handler may not be unique;
+    # imagine 5 subsectors being remapped/consolidated to 2 subsectors.
+    nonfraction_cols = [x for x in df.columns if x not in {"fraction", "from_fraction"}]
+    df = df.fillna(1, subset=["from_fraction"]).selectExpr(
+        *nonfraction_cols, "fraction*from_fraction AS fraction"
+    )
+    return df
+
+
+@track_timing(timer_stats_collector)
+def map_and_reduce_pivot_dimension(df, records, pivot_columns, operation, rename=False):
+    """Maps the pivoted dimension columns as specified by records and operation. The operation
+    is a row-wise aggregation.
+
+    This can be used to:
+        1. Map a dataset's dataframe columns per dataset-to-project mapping records.
+        2. Map a project's dataframe columns per base-to-supplemental mapping records.
+
+    Parameters
+    ----------
+    df : pyspark.sql.DataFrame
+    records : pyspark.sql.DataFrame
+        Dimension mapping records
+    pivot_columns : set
+        Column names in df that are the pivoted dimension records
+    operation : str
+        Controls how to aggregate the the mapped columns.
+    rename : bool
+        Controls whether to rename the aggregated columns to include the operation name.
+        For example, if the request is to sum the supplemental dimension 'all_electricity,'
+        the column will be renamed to 'all_electricity_sum.'
+
+    Returns
+    -------
+    tuple
+        pyspark.sql.DataFrame and sorted list of pivoted dimension columns in that table
+
+    """
+    diff = pivot_columns.difference(df.columns)
+    assert not diff, diff
+    nonvalue_cols = list(set(df.columns).difference(pivot_columns))
+    columns = set(df.columns)
+
+    records_dict = defaultdict(dict)
+    for row in records.collect():
+        if row.to_id is not None and row.from_id in columns:
+            records_dict[row.to_id][row.from_id] = row.from_fraction
+
+    to_ids = sorted(records_dict)
+    exprs = []
+    final_columns = []
+    for tid in to_ids:
+        column = f"{tid}_{operation}" if rename else tid
+        final_columns.append(column)
+        expr = [f"{from_id}*{fraction}" for from_id, fraction in records_dict[tid].items()]
+        if operation == "sum":
+            val = "+".join(expr)
+            expr = f"{val} AS {column}"
+        elif operation == "max":
+            val = ",".join(expr)
+            expr = f"greatest({val}) AS {column}"
+        elif operation == "min":
+            val = ",".join(expr)
+            expr = f"least({val}) AS {column}"
+        elif operation in ("avg", "mean"):
+            val = "(" + "+".join(expr) + f") / {len(expr)}"
+            expr = f"{val} AS {column}"
+        else:
+            raise Exception(f"Unsupported operation: {operation}")
+        exprs.append(expr)
+
+    return df.selectExpr(*nonvalue_cols, *exprs), sorted(final_columns)
```

Comment:
*exprs is not sorted but final_columns is? Maybe sort both?

Reply:
I'm not sure about sorting expressions. The table columns do get properly ordered at the end of StandardDatasetSchemaHandler.make_project_dataframe.

Reply:
This is just a sort on to_ids before the for-loop if you want to sort both expressions and final_columns